File: frontend/src/services/costForecasting.ts
/**
 * Cost Forecasting API client
 */

import { apiClient } from '@/utils/api';

export interface CostForecast {
  workflow_id: string;
  expected_queries: number;
  forecast_period_days: number;
  avg_cost_per_query: number;
  forecasted_total_cost: number;
  forecasted_daily_cost: number;
  forecasted_monthly_cost: number;
  confidence: 'low' | 'medium' | 'high' | 'none';
  sample_size: number;
  message?: string;
}

export interface CostForecastRequest {
  workflow_id: string;
  expected_queries: number;
  days?: number;
}

export interface CostTrend {
  workflow_id: string;
  period_days: number;
  daily_costs: Array<{
    date: string;
    avg_cost: number;
    query_count: number;
  }>;
  weekly_costs: Array<{
    week: string;
    avg_cost: number;
  }>;
  trend: 'increasing' | 'decreasing' | 'stable' | 'insufficient_data';
  total_queries: number;
  total_cost: number;
}

export interface CostBreakdown {
  workflow_id: string;
  period_days: number;
  breakdown: Record<string, {
    total_cost: number;
    avg_cost: number;
    count: number;
    percentage: number;
  }>;
  total_cost: number;
  total_queries: number;
}

export const costForecastingApi = {
  forecast: async (request: CostForecastRequest): Promise<CostForecast> => {
    const response = await apiClient.post<CostForecast>('/cost-forecast', {
      workflow_id: request.workflow_id,
      expected_queries: request.expected_queries,
      days: request.days || 30,
    });
    return response.data;
  },

  getTrends: async (workflowId: string, days: number = 30): Promise<CostTrend> => {
    const response = await apiClient.get<CostTrend>(`/cost-forecast/${workflowId}/trends`, {
      params: { days },
    });
    return response.data;
  },

  getBreakdown: async (workflowId: string, days: number = 30): Promise<CostBreakdown> => {
    const response = await apiClient.get<CostBreakdown>(`/cost-forecast/${workflowId}/breakdown`, {
      params: { days },
    });
    return response.data;
  },
};

File: backend/api/cost_forecasting.py
"""
API endpoints for cost forecasting.
"""

from typing import Optional
from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import BaseModel, Field

from backend.core.cost_forecasting import get_cost_forecaster
from backend.core.user_context import get_user_id_from_request
from backend.utils.logger import get_logger

logger = get_logger(__name__)

router = APIRouter(prefix="/api/v1", tags=["Cost Forecasting"])


class CostForecastRequest(BaseModel):
    """Request model for cost forecasting."""
    workflow_id: str = Field(description="Workflow ID to forecast for")
    expected_queries: int = Field(ge=1, description="Expected number of queries")
    days: int = Field(default=30, ge=1, le=365, description="Forecast period in days")


class CostForecastResponse(BaseModel):
    """Response model for cost forecast."""
    workflow_id: str
    expected_queries: int
    forecast_period_days: int
    avg_cost_per_query: float
    forecasted_total_cost: float
    forecasted_daily_cost: float
    forecasted_monthly_cost: float
    confidence: str
    sample_size: int
    message: Optional[str] = None


@router.post("/cost-forecast", response_model=CostForecastResponse)
async def forecast_cost(request: Request, forecast_request: CostForecastRequest):
    """
    Forecast cost for a workflow based on historical data.
    
    Returns cost predictions based on historical execution traces.
    """
    user_id = get_user_id_from_request(request)
    
    try:
        forecaster = get_cost_forecaster()
        forecast = forecaster.forecast_cost(
            workflow_id=forecast_request.workflow_id,
            expected_queries=forecast_request.expected_queries,
            days=forecast_request.days,
            user_id=user_id,
        )
        return CostForecastResponse(**forecast)
    except Exception as e:
        logger.error(f"Failed to forecast cost: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to forecast cost")


@router.get("/cost-forecast/{workflow_id}/trends")
async def get_cost_trends(
    request: Request,
    workflow_id: str,
    days: int = Query(default=30, ge=1, le=365),
):
    """
    Get cost trends for a workflow over time.
    
    Returns daily and weekly cost aggregations.
    """
    user_id = get_user_id_from_request(request)
    
    try:
        forecaster = get_cost_forecaster()
        trends = forecaster.analyze_cost_trends(
            workflow_id=workflow_id,
            days=days,
            user_id=user_id,
        )
        return trends
    except Exception as e:
        logger.error(f"Failed to get cost trends: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to get cost trends")


@router.get("/cost-forecast/{workflow_id}/breakdown")
async def get_cost_breakdown(
    request: Request,
    workflow_id: str,
    days: int = Query(default=30, ge=1, le=365),
):
    """
    Get cost breakdown by span type for a workflow.
    
    Shows which parts of the workflow are most expensive.
    """
    user_id = get_user_id_from_request(request)
    
    try:
        forecaster = get_cost_forecaster()
        breakdown = forecaster.get_cost_breakdown(
            workflow_id=workflow_id,
            days=days,
            user_id=user_id,
        )
        return breakdown
    except Exception as e:
        logger.error(f"Failed to get cost breakdown: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to get cost breakdown")

